## Stock Data
Now that you've had exposure to time series data, let's look at bringing stock prices into Pandas.
## Reading in Data
Your dataset can come in a variety of different formats. The most common format is the [CSV](https://en.wikipedia.org/wiki/Comma-separated_values). We'll use the "prices.csv" file as an example csv file.

In [65]:
with open('prices.csv', 'r') as file:
    prices = file.read()
    
print(prices)

ABC,2017-09-05,163.09,164.24,160.21,162.63,29417590.0,162.49,29414672.0
ABC,2017-09-06,162.85,162.46,159.99,161.13,21131267.0,162.44,21169319.0
ABC,2017-09-07,162.11,162.7,160.65,161.26,21722502.0,161.46,21719856.0
ABC,2017-09-08,160.41,160.89,159.31,158.05,28311012.0,158.26,28305810.0
ABC,2017-09-11,161.09,162.14,159.54,161.29,31075573.0,160.97,31163734.0
ABC,2017-09-12,162.54,164.61,159.52,161.09,70921229.0,160.62,71097150.0
ABC,2017-09-13,160.01,160.51,158.22,159.29,44580353.0,159.07,44260255.0
EFG,2017-09-05,154.45,154.69,153.17,154.52,1270203.0,153.58,1270679.0
EFG,2017-09-06,155.03,155.14,153.89,154.45,1195987.0,154.06,1196107.0
EFG,2017-09-07,154.73,155.36,153.6,155.68,1420730.0,155.6,1409098.0
EFG,2017-09-08,156.01,155.91,154.17,155.86,1438929.0,156.08,1445338.0
EFG,2017-09-11,157.07,157.71,155.93,157.17,1608840.0,156.7,1610357.0
EFG,2017-09-12,155.98,156.72,154.28,156.71,1692197.0,156.86,1687819.0
EFG,2017-09-13,156.4,157.07,155.68,155.54,1211779.0,155.85,1210716.0
XYZ,2017-09

The data provider will provide you with information for each field in the CSV. This csv has the fields ticker, date, open, high, low, close, volume, adj_close, adj_volume in that order. That means, the first line in the CSV has the following data:

- ticker: ABC
- date: 2017-09-05
- open: 163.09
- high: 164.24
- low: 160.21
- close: 162.63
- volume: 29417590.0
- adj_close: 162.49
- adj_volume: 29414672.0

Let's move this data into a DataFrame. For this, we'll need to use the [`pd.read_csv`](https://pandas.pydata.org/pandas-docs/version/0.21.1/generated/pandas.read_csv.html) function. This allows you generate a DataFrame from CSV data.

In [66]:
import pandas as pd

price_df = pd.read_csv('prices.csv')

price_df.head()

ABC  2017-09-05  163.09  164.24  160.21  162.63  29417590.0  162.49  \
0  ABC  2017-09-06  162.85  162.46  159.99  161.13  21131267.0  162.44   
1  ABC  2017-09-07  162.11  162.70  160.65  161.26  21722502.0  161.46   
2  ABC  2017-09-08  160.41  160.89  159.31  158.05  28311012.0  158.26   
3  ABC  2017-09-11  161.09  162.14  159.54  161.29  31075573.0  160.97   
4  ABC  2017-09-12  162.54  164.61  159.52  161.09  70921229.0  160.62   

   29414672.0  
0  21169319.0  
1  21719856.0  
2  28305810.0  
3  31163734.0  
4  71097150.0

That generated a DataFrame using the CSV, but assumed the first row contains the field names. We'll have to supply the function's parameter `names` with a list of fiels names.

In [67]:
price_df = pd.read_csv('prices.csv', names=['ticker', 'date', 'open', 'high', 'low',
                                             'close', 'volume', 'adj_close', 'adj_volume'])

price_df.head()

ticker        date    open    high     low   close      volume  adj_close  \
0    ABC  2017-09-05  163.09  164.24  160.21  162.63  29417590.0     162.49   
1    ABC  2017-09-06  162.85  162.46  159.99  161.13  21131267.0     162.44   
2    ABC  2017-09-07  162.11  162.70  160.65  161.26  21722502.0     161.46   
3    ABC  2017-09-08  160.41  160.89  159.31  158.05  28311012.0     158.26   
4    ABC  2017-09-11  161.09  162.14  159.54  161.29  31075573.0     160.97   

   adj_volume  
0  29414672.0  
1  21169319.0  
2  21719856.0  
3  28305810.0  
4  31163734.0

## DataFrame Calculations
Now that we have the data in a DataFrame, we can start to do calculations on it. Let's find out the median value for each stock using the [`DataFrame.median`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.median.html) function.

In [68]:
price_df.median()

open              155.98
high              155.91
low               154.17
close             155.68
volume        3730110.00
adj_close         155.85
adj_volume    3725435.00
dtype: float64

That's not right. Those are the median values for the whole stock universe. We'll use the [`DataFrame.groupby`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.groupby.html) function to get mean for each stock.

In [69]:
price_df.groupby('ticker').median()

open    high     low   close      volume  adj_close  adj_volume
ticker                                                                   
ABC     162.11  162.46  159.54  161.13  29417590.0     160.97  29414672.0
EFG     155.98  155.91  154.17  155.68   1420730.0     155.85   1409098.0
XYZ      60.36   60.66   58.80   60.33   3730110.0      60.46   3725435.0

That's what we're looking for! However, we don't want to run the `groupby` function each time we make an operation. We could save the GroupBy object by doing `price_df_ticker_groups = price_df.groupby('ticker')`. This limits us to the operations of GroupBy objects. There's the [`GroupBy.apply`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.core.groupby.GroupBy.apply.html), but then we lose out on performance. The true problem is the way the data is represented.

In [90]:
price_df.iloc[:16]

ticker        date         open         high          low        close  \
0     ABC  2017-09-05 163.09000000 164.24000000 160.21000000 162.63000000   
1     ABC  2017-09-06 162.85000000 162.46000000 159.99000000 161.13000000   
2     ABC  2017-09-07 162.11000000 162.70000000 160.65000000 161.26000000   
3     ABC  2017-09-08 160.41000000 160.89000000 159.31000000 158.05000000   
4     ABC  2017-09-11 161.09000000 162.14000000 159.54000000 161.29000000   
5     ABC  2017-09-12 162.54000000 164.61000000 159.52000000 161.09000000   
6     ABC  2017-09-13 160.01000000 160.51000000 158.22000000 159.29000000   
7     EFG  2017-09-05 154.45000000 154.69000000 153.17000000 154.52000000   
8     EFG  2017-09-06 155.03000000 155.14000000 153.89000000 154.45000000   
9     EFG  2017-09-07 154.73000000 155.36000000 153.60000000 155.68000000   
10    EFG  2017-09-08 156.01000000 155.91000000 154.17000000 155.86000000   
11    EFG  2017-09-11 157.07000000 157.71000000 155.93000000 157.17000000   
12    EFG  2017-09-12 155.98000000 156.72000000 154.28000000 156.71000000   
13    EFG  2017-09-13 156.40000000 157.07000000 155.68000000 155.54000000   
14    XYZ  2017-09-05  63.90000000  64.51000000  63.13000000  63.95000000   
15    XYZ  2017-09-06  63.85000000  63.65000000  61.72000000  62.23000000   

              volume    adj_close        adj_volume  
0  29417590.00000000 162.49000000 29414672.00000000  
1  21131267.00000000 162.44000000 21169319.00000000  
2  21722502.00000000 161.46000000 21719856.00000000  
3  28311012.00000000 158.26000000 28305810.00000000  
4  31075573.00000000 160.97000000 31163734.00000000  
5  70921229.00000000 160.62000000 71097150.00000000  
6  44580353.00000000 159.07000000 44260255.00000000  
7   1270203.00000000 153.58000000  1270679.00000000  
8   1195987.00000000 154.06000000  1196107.00000000  
9   1420730.00000000 155.60000000  1409098.00000000  
10  1438929.00000000 156.08000000  1445338.00000000  
11  1608840.00000000 156.70000000  1610357.00000000  
12  1692197.00000000 156.86000000  1687819.00000000  
13  1211779.00000000 155.85000000  1210716.00000000  
14  1738651.00000000  63.33000000  1733249.00000000  
15  3730110.00000000  61.95000000  3725435.00000000

In [122]:
price_df['date']=pd.to_datetime(price_df['date'])

In [134]:
df = price_df.copy()
df.set_index(['date', 'ticker'],inplace=True)
df.sort_index(inplace=True)

In [135]:
df

open         high          low        close  \
date       ticker                                                       
2017-09-05 ABC    163.09000000 164.24000000 160.21000000 162.63000000   
           EFG    154.45000000 154.69000000 153.17000000 154.52000000   
           XYZ     63.90000000  64.51000000  63.13000000  63.95000000   
2017-09-06 ABC    162.85000000 162.46000000 159.99000000 161.13000000   
           EFG    155.03000000 155.14000000 153.89000000 154.45000000   
           XYZ     63.85000000  63.65000000  61.72000000  62.23000000   
2017-09-07 ABC    162.11000000 162.70000000 160.65000000 161.26000000   
           EFG    154.73000000 155.36000000 153.60000000 155.68000000   
           XYZ     61.97000000  61.93000000  59.47000000  60.46000000   
2017-09-08 ABC    160.41000000 160.89000000 159.31000000 158.05000000   
           EFG    156.01000000 155.91000000 154.17000000 155.86000000   
           XYZ     60.36000000  60.45000000  58.51000000  59.35000000   
2017-09-11 ABC    161.09000000 162.14000000 159.54000000 161.29000000   
           EFG    157.07000000 157.71000000 155.93000000 157.17000000   
           XYZ     60.04000000  59.92000000  57.68000000  58.24000000   
2017-09-12 ABC    162.54000000 164.61000000 159.52000000 161.09000000   
           EFG    155.98000000 156.72000000 154.28000000 156.71000000   
           XYZ     58.19000000  59.29000000  57.89000000  58.71000000   
2017-09-13 ABC    160.01000000 160.51000000 158.22000000 159.29000000   
           EFG    156.40000000 157.07000000 155.68000000 155.54000000   
           XYZ     59.01000000  60.66000000  58.80000000  60.33000000   

                             volume    adj_close        adj_volume  
date       ticker                                                   
2017-09-05 ABC    29417590.00000000 162.49000000 29414672.00000000  
           EFG     1270203.00000000 153.58000000  1270679.00000000  
           XYZ     1738651.00000000  63.33000000  1733249.00000000  
2017-09-06 ABC    21131267.00000000 162.44000000 21169319.00000000  
           EFG     1195987.00000000 154.06000000  1196107.00000000  
           XYZ     3730110.00000000  61.95000000  3725435.00000000  
2017-09-07 ABC    21722502.00000000 161.46000000 21719856.00000000  
           EFG     1420730.00000000 155.60000000  1409098.00000000  
           XYZ     6166046.00000000  60.64000000  6191712.00000000  
2017-09-08 ABC    28311012.00000000 158.26000000 28305810.00000000  
           EFG     1438929.00000000 156.08000000  1445338.00000000  
           XYZ     5173590.00000000  59.40000000  5174940.00000000  
2017-09-11 ABC    31075573.00000000 160.97000000 31163734.00000000  
           EFG     1608840.00000000 156.70000000  1610357.00000000  
           XYZ     5003322.00000000  58.02000000  5001118.00000000  
2017-09-12 ABC    70921229.00000000 160.62000000 71097150.00000000  
           EFG     1692197.00000000 156.86000000  1687819.00000000  
           XYZ     3633446.00000000  58.96000000  3635132.00000000  
2017-09-13 ABC    44580353.00000000 159.07000000 44260255.00000000  
           EFG     1211779.00000000 155.85000000  1210716.00000000  
           XYZ     3571591.00000000  60.46000000  3583560.00000000

Can you spot our problem? Take a moment to see if you can find it.

The problem is between lines [6, 7] and  [13, 14]

In [71]:
price_df.iloc[[6, 7, 13, 14]]

ticker        date    open    high     low   close      volume  adj_close  \
6     ABC  2017-09-13  160.01  160.51  158.22  159.29  44580353.0     159.07   
7     EFG  2017-09-05  154.45  154.69  153.17  154.52   1270203.0     153.58   
13    EFG  2017-09-13  156.40  157.07  155.68  155.54   1211779.0     155.85   
14    XYZ  2017-09-05   63.90   64.51   63.13   63.95   1738651.0      63.33   

    adj_volume  
6   44260255.0  
7    1270679.0  
13   1210716.0  
14   1733249.0

Data for all the tickers are stacked. We're representing 3 dimensional data in 2 dimensions. This was solved using Panda's Panels, which is [deprecated](https://pandas.pydata.org/pandas-docs/version/0.21.0/dsintro.html#deprecate-panel). The Pandas documentation recommends we use either [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) or [xarray](http://xarray.pydata.org/en/stable/). [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) still doesn't solve our problem, since the data is still represented in 2 dimensions. [xarray](http://xarray.pydata.org/en/stable/) is able to store 3 dimensional data, but Finance uses Pandas, so we'll stick with this library. After you finish this program, I recommend you check out [xarray](http://xarray.pydata.org/en/stable/).

So, how do we use our 3-dimensional data with Pandas? We can split each 3rd dimension into it's own 2 dimension DataFrame. Let's take this array as an example:
```
[
    [
        [ 0,  1],
        [ 2,  3],
        [ 4,  5]
    ],[
        [ 6,  7],
        [ 8,  9],
        [10, 11]
    ],[
        [12, 13],
        [14, 15],
        [16, 17]
    ],[
        [18, 19],
        [20, 21],
        [22, 23]
    ]
]    
```
We want to split it into these two 2d arrays:
```
[
    [0, 2, 4],
    [6, 8, 10],
    [12, 14, 16],
    [18, 20, 22]
] 
[
    [1, 3, 5],
    [7, 8, 11],
    [13, 15, 17],
    [19, 21, 23]
] 
```
In our case, our third dimensions are "open", "high", "low", "close", "volume", "adj_close", and  "adj_volume". We'll use the [`DataFrame.pivot`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.pivot.html) function to generate these DataFrames.

In [77]:
open_prices = price_df.pivot(index='date', columns='ticker', values='open')
high_prices = price_df.pivot(index='date', columns='ticker', values='high')
low_prices = price_df.pivot(index='date', columns='ticker', values='low')
close_prices = price_df.pivot(index='date', columns='ticker', values='close')
volume = price_df.pivot(index='date', columns='ticker', values='volume')
adj_close_prices = price_df.pivot(index='date', columns='ticker', values='adj_close')
adj_volume = price_df.pivot(index='date', columns='ticker', values='adj_volume')

close_prices

ticker              ABC          EFG         XYZ
date                                            
2017-09-05 162.63000000 154.52000000 63.95000000
2017-09-06 161.13000000 154.45000000 62.23000000
2017-09-07 161.26000000 155.68000000 60.46000000
2017-09-08 158.05000000 155.86000000 59.35000000
2017-09-11 161.29000000 157.17000000 58.24000000
2017-09-12 161.09000000 156.71000000 58.71000000
2017-09-13 159.29000000 155.54000000 60.33000000

That gives you DataFrames for all the open, high low, etc.. Now, what we have been waiting for.. The mean for each ticker.

In [73]:
open_prices.mean()

ticker
ABC    161.728571
EFG    155.667143
XYZ     61.045714
dtype: float64

We can also get the mean for each date by doing a transpose.

In [74]:
open_prices.T.mean()

date
2017-09-05    127.146667
2017-09-06    127.243333
2017-09-07    126.270000
2017-09-08    125.593333
2017-09-11    126.066667
2017-09-12    125.570000
2017-09-13    125.140000
dtype: float64

It doesn't matter whether date is the index and tickers are the colums or the other way around. It's always a transpose away. Since we're going to do a lot of operations across dates, we will stick with date as the index and tickers as the colums throughtout this program.
## Quiz
Let's see if you can apply what you learned. Implment the `csv_to_close` function to take in a filepath, `csv_filename`, and output the close 2d array. You can assume the CSV file used by `csv_to_close` has the same field names as "prices.csv" and in the same order.

To help with your implemention of quizzes, we provide you with unit tests to test your function implemention. For this quiz, we'll be using the function `test_csv_to_close` in the `quiz_tests` module to test `csv_to_close`.

In [76]:
import quiz_tests


def csv_to_close(csv_filepath, field_names):
    """Reads in data from a csv file and produces a DataFrame with close data.
    
    Parameters
    ----------
    csv_filepath : str
        The name of the csv file to read
    field_names : list of str
        The field names of the field in the csv file

    Returns
    -------
    close : DataFrame
        Close prices for each ticker and date
    """
    result = pd.read_csv(csv_filepath, names=field_names).pivot(index='date', columns='ticker', values='close')
    # TODO: Implement Function
    
    return result


quiz_tests.test_csv_to_close(csv_to_close)

Tests Passed


## Quiz Solution
If you're having trouble, you can check out the quiz solution [here](stock_data_solution.ipynb).